feat: DPMMA-1572 Clone campaign events

patrykgruszka opened the pull request:
<!-- ## Which branch should I use for my PR?

Assuming that:

a = current major release
b = current minor release
c = future major release

* a.x for any features and enhancements (e.g. 5.x)
* a.b for any bug fixes (e.g. 4.4, 5.1)
* c.x for any features, enhancements or bug fixes with backward compatibility breaking changes (e.g. 5.x) -->

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [ N ]
| New feature/enhancement? (use the a.x branch)      | [ Y ]
| Deprecations?                          | [ N ]
| BC breaks? (use the c.x branch)        | [ N ]
| Automated tests included?              | [ Y ] <!-- All PRs must maintain or improve code coverage -->
| Related user documentation PR URL      | mautic/mautic-documentation#... <!-- required for new features -->
| Related developer documentation PR URL | mautic/developer-documentation#... <!-- required for developer-facing changes -->
| Issue(s) addressed                     | Fixes #... <!-- prefix each issue number with "Fixes #", no need to create an issue if none exists, explain below instead -->

<!--
Additionally (see https://contribute.mautic.org/contributing-to-mautic/developer/code/pull-requests#work-on-your-pull-request):
 - Always add tests and ensure they pass.
 - Bug fixes must be submitted against the lowest maintained branch where they apply
   (lowest branches are regularly merged to upper ones so they get the fixes too.)
 - Features and deprecations must be submitted against the "4.x" branch.
-->

#### Description:
This PR adds possibility to clone (copy/paste) campaign events. 
Cloning events between campaigns is supported with this feature.

![image](https://github.com/mautic/mautic/assets/8580942/86617085-39a0-4e4f-be76-f4d47272181d)
![image](https://github.com/mautic/mautic/assets/8580942/b1492b07-06df-4909-87ab-ac5b20535d0d)


<!--
Please write a short README for your feature/bugfix. This will help people understand your PR and what it aims to do. If you are fixing a bug and if there is no linked issue already, please provide steps to reproduce the issue here.
-->

#### Steps to test this PR:

<!--
This part is really important. If you want your PR to be merged, take the time to write very clear, annotated and step by step test instructions. Do not assume any previous knowledge - testers may not be developers.
-->
1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Create a campaign with some events
3. Clone some events and paste them into other branches or other campaigns
4. Run the campaign and check that it works properly
<!--
If you have any deprecations, list them here along with the new alternative.
If you have any backwards compatibility breaks, list them here.
-->

patrykgruszka committed:
feat: [DPMMA-1572] clone campaign events draft

patrykgruszka committed:
feat: [DPMMA-1572] paste cloned events

patrykgruszka committed:
feat: [DPMMA-1572] use localStorage for cloned event sync

patrykgruszka committed:
feat: [DPMMA-1572] event clone styles

patrykgruszka committed:
fix: [DPMMA-1572] campaign event add/edit functional tests

patrykgruszka committed:
fix: [DPMMA-1572] EventController refactor

patrykgruszka committed:
fix: [DPMMA-1572] ux improvments, refactor

patrykgruszka committed:
fix: [DPMMA-1572] fix event edit and ux improvements

patrykgruszka committed:
fix: [DPMMA-1572] event clone functional tests

Robmobius reviewed (approved):
Seems to work. Created a campaign, created a tagging action, cloned the action and ran the campaign and it worked as expected.

jos0405 commented:
Tried different events, all works.
Campaign is working properly after using cloned events.

![image](https://github.com/mautic/mautic/assets/4246909/2d82a778-da39-4771-95cc-aa96c78e5d45)

jos0405 commented:
I noticed UI bugs, but probably unrelated:
![image](https://github.com/mautic/mautic/assets/4246909/fa9cf251-f22a-491c-819f-0721baaf22d3)

tomekkowalczyk reviewed (approved):
@patrykgruszka I tested feature. Seems to work. Good job.

escopecz commented on app/config/local.php.ses:
```diff
@@ -0,0 +1,782 @@
+<?php
```
This was probably pushed by accident, right? Please remove it and check if you haven't published some secrets. If so, reset those just to be sure.

patrykgruszka replied:
Yes, this was by accident. Thank you

kuzmany commented:
Anyway, do not see so many benefits of the PR.

It can clone events
You can clone 1 event at a time. Is it reset after cloning from memory? If I want to clone it 2 times I need to clone it 2 times?
It's stored just to a session, then it's lost.

This clone should be done directly in the builder without so many steps.  Just clone and attach to any path.

Or I imagine creating some storage to use predefined events from memory anytime to any campaign.

patrykgruszka commented:
@kuzmany it works just like copy/paste, but the storage is cleared when the object is used. I can easily change the behavior and do not clear the storage, so the element could be reused. There also could be added a functionality to clone the whole branch in the future. This one would require more work. 

> This clone should be done directly in the builder without so many steps. Just clone and attach to any path.

I don't get it. The whole process is just 3 clicks and it's done right in the builder.

This feature has been requested by users within our company and the Mautic forum community.

patrykgruszka commented:
@escopecz, @kuzmany I've rebased the code onto 5.1. Is it ready to go as is, or do you have any suggestions for changes?

escopecz commented on app/bundles/CampaignBundle/Assets/js/campaign.js:
```diff
@@ -1725,6 +1764,9 @@ Mautic.campaignBuilderUpdateEventList = function (groups, hidden, view, active,
         }
     });
 
+    mQuery('#CampaignEventPanelGroups').removeClass('groups-enabled-1 groups-enabled-2 groups-enabled-3')
+    mQuery('#CampaignEventPanelGroups').addClass('groups-enabled-' + groups.length)
```
Can you explain what this is for? I can see the `groups-enabled-2` has some CSS rule, but nothing else. What happens with `groups-enabled-4`?

patrykgruszka replied:
It's the number of available event types to select in this window:
![image](https://github.com/mautic/mautic/assets/8580942/4b19e097-ea57-45e5-b2b0-4da80bbe3884)
For example, after the decision event, you can only choose action or condition, so the window will have the class `groups-enabled-2` so that it can be properly styled for that size. The 3 groups are max at this moment, I'm clearing here all possible group classes before adding the current one.

escopecz commented:
I noticed that the decision list contains all decisions. Previously, some decisions were available only after certain actions.
Before:
<img width="463" alt="Screenshot 2024-01-08 at 14 12 29" src="https://github.com/mautic/mautic/assets/1235442/51c6e14c-2974-41fc-9e4c-eb0f854e396d">

After
<img width="485" alt="Screenshot 2024-01-08 at 14 15 39" src="https://github.com/mautic/mautic/assets/1235442/2cffcd12-3702-4132-b77d-6dbc09b7c46c">

And even though I cannot select a decision after a decision the clone feature allows me to do that:
<img width="481" alt="Screenshot 2024-01-08 at 14 16 14" src="https://github.com/mautic/mautic/assets/1235442/d9985971-7e13-414e-be79-b5543b5ecba6">

patrykgruszka committed:
fix: [DPMMA-1572] undefined clone details

patrykgruszka committed:
feat: [DPMMA-1572] clone campaign node restrictions

patrykgruszka committed:
Merge branch '5.x' into DPMMA-1572_copy-campaign-events

patrykgruszka committed:
fix: [DPMMA-1572] event clone meessage

patrykgruszka commented:
@escopecz good catch, I've fixed the problem with filtering the list and added restrictions to the clone actions. 

There is still some inconsistency with the nodes connection validation, but it existed before this PR. For example:
- **not allowed:** Start campaign action -> Opens email decision 
- **allowed:** Start campaign action -> Any condition -> Opens email decision

escopecz committed:
Merge branch '5.x' into DPMMA-1572_copy-campaign-events

code5rick commented:
Loved the clone feature, awesome work! I would only suggest some refinements in terms of UI:

**Campaign List Diagram (hover):**
Actual:
![image](https://github.com/mautic/mautic/assets/131262825/f260e07e-c2a8-4cc6-97b1-5c4470e26d96)

Proposal:
![image](https://github.com/mautic/mautic/assets/131262825/aed39efc-a372-4ef9-a8d4-63a06b61188d)

- Add a hover effect similar to the close button for consistency. The edit and clone buttons currently lack this feature.
- Remove the border.
- Increase the size of the circle to 24x24px.
- Enlarge the icon to a font size of 12px.
- Change the color of the clone circle to differentiate it. This allows users to associate colors with actions, facilitating easier recognition and distinction of each action.
- Adjust the position of the .campaign-event-buttons class to top: -14px; right: -12px; so that the text inside the card can be more easily readable.

-----

**Campaign event panel groups:**
Actual:
![image](https://github.com/mautic/mautic/assets/131262825/81dda421-d137-475c-b581-a81827456d34)

Proposal:
![image](https://github.com/mautic/mautic/assets/131262825/9f4f1310-3cb4-4896-95bf-bc25d3aab01f)

- While I appreciate the current design of the clone container, it should maintain consistency with the other containers.
- To enhance recognition of whether it's a decision, action, or condition, the event type should be positioned in front of the event name as a labeled tag, color-coded accordingly.
- (Optional) Remove clone button. If a user accidentally clones an event, the clone section will always appear below the container. Allowing users to remove the clone would provide added flexibility.
- Use the same gray color as the button for the panel title of the clone.

kuzmany commented:
I suggest adding UX changes in a separate PR

patrykgruszka commented:
So, should we proceed with the proposal or merge it as it is and discuss the UI design later? It's a good point to keep the container consistent, but on the other hand the whole window is getting bigger with the new suggestion. It may be an issue on smaller screens. Also, I'm not sure about the description "A decision is made..." - should it be just a copy of the description for decision/action/condition? 

I'm good with making the changes, just want to double-check everyone's okay with them first.

andersonjeccel reviewed (approved):
I think it's better to leave as it is now, focusing on making it to be merged

patrykgruszka commented:
@code5rick, @kuzmany, @andersonjeccel thank you for sharing your perspective.

@escopecz do you agree with merging this PR as it stands?

escopecz reviewed (dismissed):
Thanks! 👍

escopecz committed:
Merge branch '5.x' into DPMMA-1572_copy-campaign-events

RCheesley commented:
Hi @patrykgruszka - sorry but here's another PR without the docs PR to accompany it.

I think this will best sit on this page: https://docs.mautic.org/en/5.x/campaigns/campaign_builder.html

CC @mautic/education-team as this is required for the 5.1 release.

putzwasser commented:
@patrykgruszka Just to verify: Does a cloned event depend on its _source_?

I.e., I create an event. I clone and paste that event. Now I change the source event without touching the pasted clone:

Does the cloned event reflect the changes I made to the source? Or is the cloned event fully independent (and I'd need to make the same change to both events)?

patrykgruszka commented:
@putzwasser it's fully independent after clone.
